[AArch64] Expand vector ops when NEON and SVE are unavailable.

sdesmalen-arm opened the pull request:
Unlike `+noneon` we must assume that vector types are available, i.e. it is
valid to pass/return vector arguments to and from functions. However, the
compiler must make sure to scalarize any vector operations.

sdesmalen-arm committed:
[AArch64] Expand vector ops when NEON and SVE are unavailable.

Unlike `+noneon` we must assume that vector types are available, i.e. it is
valid to pass/return vector arguments to and from functions. However, the
compiler must make sure to scalarize any vector operations.

sdesmalen-arm commented:
This PR is currently still based off #90723 but I'll rebase once that PR lands.

The patch to review is: https://github.com/llvm/llvm-project/pull/90833/commits/2270ec14ae56733c0b65064c109f2ece6215829f

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -378,6 +378,27 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
     addQRTypeForNEON(MVT::v8f16);
     addQRTypeForNEON(MVT::v8bf16);
+  } else if (Subtarget->hasNEON() || Subtarget->useSVEForFixedLengthVectors()) {
+    addRegisterClass(MVT::v16i8, &AArch64::FPR8RegClass);
+    addRegisterClass(MVT::v8i16, &AArch64::FPR16RegClass);
+
+    addRegisterClass(MVT::v2f32, &AArch64::FPR64RegClass);
```
I don't really like the duplication here.  What about passing `isNeonAvailable()` into `addDRTypeForNEON` and `addQRTypeForNEON` to toggle the call to `addTypeForNEON`?

sdesmalen-arm replied:
I've moved this into (the renamed functions) `addQRType` and `addDRType` now.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1328,6 +1349,24 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
     // FADDP custom lowering
     for (MVT VT : { MVT::v16f16, MVT::v8f32, MVT::v4f64 })
       setOperationAction(ISD::FADD, VT, Custom);
+  } else {
```
Alternatively should this be `else if (!Subtarget->useSVEForFixedLengthVectors)`? Just thinking there's several instances of "set everything to Expand" going on.

sdesmalen-arm committed:
Address review comments

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -3731,8 +3731,10 @@ bool SelectionDAGLegalize::ExpandNode(SDNode *Node) {
     EVT VT = Node->getValueType(0);
-    assert(TLI.isOperationLegalOrCustom(ISD::ADD, VT) &&
-           TLI.isOperationLegalOrCustom(ISD::XOR, VT) &&
+    assert((VT.isFixedLengthVector() || // fixed length ADD can be expanded to
```
This change looks weird.  The assert is checking whether it is safe to create `ADD` and `XOR` nodes, which you're bypassing?  What does the failing test case look like?  Part of me thinks the original assert should be an `if` statement, but then I'm also wondering why `VectorLegalizer::LegalizeOp` hasn't already unrolled the operation.

sdesmalen-arm replied:
That's right. The way the code now works (if you'd ignore the assert) is that it expands a `sub` into a `add` + `xor`. If those are vector operations that also require expansion, they can be expanded further (in a subsequent step) into scalar operations. The assert is trying to match the way the code currently works, which allows the code to be successfully compiled. I figured updating the assert was the most sensible thing to do for now, given that it's not really a common use-case to optimise for.

paulwalker-arm replied:
To me the assert exists to ensure optimal code, which is to say it doesn't really make sense to convert the sub to an inverted add unless both those operations are in someway legal.  When not the better result would be to just expand the sub.  Which I guess makes the assert a bear trap for the first person that needs the support.

sdesmalen-arm committed:
Restore assert

Instead disable the 'performAddSubIntoVectorOp' which turned a scalar i64 sub
back into a v1i64 sub.

I tried disabling the combine before (operation) legalization, but that didn't work.
So instead, I've just disabled it if the ISD::SUB needs Expand for v1i64.

paulwalker-arm reviewed (approved):
Given the size of the generated code I think we might start to question the usefulness of some of the tests (e.g. testing v32i8 variants).
